### PR TITLE
H219: Fresh H185 seed + within-recipe LERP — test mode connectivity

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis H219: Fresh H185 seed + within-recipe LERP — test mode connectivity

**Advisor branch:** tay  
**Student:** thorfinn  
**Hypothesis ID:** H219

---

### Background

Finding R (H211) showed that cross-recipe LERP (linearly interpolating between checkpoints trained with different hyperparameter recipes) is catastrophic — the models are in different loss basins. However, Finding T/U (H207/H214) showed that the sub-alpha basin radius around the SOTA checkpoint is very small (< 0.005), meaning small perturbations already leave the good basin.

**The question this experiment answers:** does a *different random seed* trained with the *same recipe* (same hyperparameters as H185) land in a connected basin? If two H185-recipe checkpoints are in the same basin (i.e., the straight-line interpolation between them stays in the same loss landscape valley), then LERP between them is valid and we can use weight averaging to find a better point in that basin.

This tests mode connectivity — a concept from the loss landscape literature (Garipov et al. 2018) — within the H185 recipe.

### Hypothesis

> Two independent H185-recipe runs (same hyperparameters, different random seeds) will lie in linearly connected loss basins, and LERP between their EP13 checkpoints at α=0.5 will produce a model with `val_abupt` at or better than the best individual checkpoint.

### Experiment

**Step 1 — Train a fresh H185 seed:**

```bash
torchrun --standalone --nproc-per-node=8 target/train.py \
  --seed 42 \
  --wandb_group h219-within-recipe-lerp \
  $SENPAI_EXTRA_ARGS
```

(Use the same hyperparameters as H185 / `yw2a5dyl`. If there is a `--config` flag pointing to the H185 config, use that. If not, match all flags manually from the H185 W&B run config.)

**Step 2 — LERP evaluation:**

Once training is complete, use `target/interp_eval.py` to interpolate between the fresh seed EP13 checkpoint and H185 `yw2a5dyl` EP13 EMA checkpoint at α ∈ {0.0, 0.25, 0.5, 0.75, 1.0}:

```bash
python target/interp_eval.py \
  --run_a yw2a5dyl \
  --run_b <your_new_run_id> \
  --alphas 0.0 0.25 0.5 0.75 1.0 \
  --wandb_group h219-within-recipe-lerp
```

### Current SOTA baseline (target to beat)

| Metric | Value |
|---|---|
| val_abupt | 5.9755 |
| test_abupt | 5.8221 |
| test_VP | 3.421 |
| test_SP | 3.577 |
| test_WSS | 6.727 |

**SOTA gate:** `val_abupt < 5.9755 AND test_abupt < 5.8221`

### What to watch

- Is there a "valley" (val_abupt monotonically bad at α=0.5, then recovering) or a "ridge" (catastrophic collapse like cross-recipe)?
- Does any α produce val_abupt < 5.9755? That would be a new SOTA from weight averaging alone.
- If connectivity is confirmed (smooth interpolation curve with no barrier), report that explicitly — it opens the door to SWA and model soup strategies within the same recipe.

### Expected outcome

If the runs are in the same basin: smooth val_abupt curve across α, possibly with a minimum near α=0.5. This enables same-recipe SWA.

If the runs are in different basins: val_abupt barrier / catastrophic collapse at intermediate α values, like cross-recipe LERP. This means even same-recipe seeds diverge and weight averaging is globally off the table.

### Instructions to student

1. Train a fresh H185-recipe run with `--seed 42` (or any new seed different from H185's).
2. Run `interp_eval.py` between your new run and `yw2a5dyl` at α ∈ {0.0, 0.25, 0.5, 0.75, 1.0}.
3. Report the val_abupt curve across all α values in a PR comment (table format).
4. Note whether the interpolation is smooth (basin connectivity) or shows a barrier.
5. If any α achieves val_abupt < 5.9755, run TTA eval: `python target/eval_tta_h209.py --wandb_run_id <best_alpha_run_id>`
6. Post the terminal `SENPAI-RESULT` marker. Use the best val_abupt found across all α values.

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt","value":<number>},"test_metric":{"name":"test_abupt","value":<number>}}
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT thorfinn: budget/scope sanity check before launching the train arm.

**Budget math:**
- H185 `yw2a5dyl` actual `total_train_minutes` = **874.4 min** (~14.6h) for 13 epochs on 8 GPUs (epoch_time_s≈3099 ≈51 min/epoch).
- My `SENPAI_TIMEOUT_MINUTES=360` (6h) is the hard student cap — instructions explicitly say "hard upper bounds, not targets".
- Even with val budget collapsed, I can only complete ~5 epochs of the H185 recipe in 6h, not EP13.

**Code state on `tay`:**
- `--mirror-augmentation` flag (used by H185 launch script) is **not in `tay`'s train.py** — the H185 commit `c7abd18` lives on `thorfinn/h171-plateau-exact`, never merged. I would need to cherry-pick it.
- No `--seed` flag exists in `tay`'s train.py — also needs to be added.
- No `interp_eval.py` exists — I would write it from scratch (load both checkpoints, LERP at α, run `evaluate_split` against val_surface).

**Mode connectivity at non-EP13 is not equivalent:** an EP5 fresh-seed checkpoint (truncated cosine T_max=5) vs `yw2a5dyl` EP13 (full cosine T_max=13) are at different points in their LR schedules — the comparison is not a clean within-recipe LERP test.

**Questions:**
1. **Budget**: am I authorized to set `SENPAI_TIMEOUT_MINUTES=900` for this run (matches advisor's H185 precedent) to actually reach EP13? Or is the 360 cap hard?
2. **Alternative**: if 6h is hard, would you prefer a truncated EP5 within-recipe LERP (matched recipes both with T_max=5 cosine), or should I abort H219 and you reassign?
3. **`interp_eval.py`**: confirm I should build it (no existing impl). Plan: load checkpoint A + checkpoint B from W&B artifacts, parameter-wise w_α = (1-α)·θ_A + α·θ_B, build model from saved config, run rank-0 `evaluate_split` against the val_surface loader, log per-α metrics under `lerp/alpha_<x>/val_*`.

Pinging for guidance — preparing the code changes (`--seed` add, mirror-aug cherry-pick, interp_eval.py) in parallel so I can launch immediately on your decision.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1396 because ADVISOR — CLOSING WITH APOLOGY. The Round 4 assignment had two systemic errors that I (advisor) own:

**Error 1: Recipe spec did not match actual yw2a5dyl config.** I cited the H185 recipe as `tau_y=3.0, tau_z=2.0, --mirror-augment-p 0.25, --grad-clip 1.0, --lr-schedule cosine, --enable-residual-positions, --save-best-checkpoint`. Verified from W&B config on yw2a5dyl, the ACTUAL recipe is `tau_y=1.3, tau_z=1.67, mirror_augmentation=True (boolean), --grad-clip-norm 0.5, --lr-cosine-t-max 13`. The other flags don't exist on tay. Student catches were correct.

**Error 2: Budget infeasibility.** yw2a5dyl actual `total_train_minutes=874.4` (~14.6h on 8 GPUs). The 6h SENPAI_TIMEOUT_MINUTES cap CANNOT fit a 13-epoch H185-recipe retrain. The Round 4 assignment was incompatible with the budget cap from the start.

**Mirror augmentation also not on tay:** verified via grep — `--mirror-augmentation` lives on `askeladd/h148-mirror-augmentation` and `fern/h183-mirror-aug-tau-y-3p0-compound` branches, never merged. Reproducing H185 recipe requires cherry-pick PR first.

BASELINE.md has been corrected on tay (commit pending).

**Reassignment incoming**: H230 — TTA-mirror applied to H183 (yutg7) and H190 (yk6gnn) cross-checkpoints — extends Finding Q to N=5 (eval-only, ~45min)

Thank you for catching the recipe discrepancies before launching. Apologies for the wasted setup time..
